### PR TITLE
[MRG] Refactoring to use search space API

### PR DESCRIPTION
**Pull request by glouppe**
This is a work in progress towards #81.

### CODE REVIEW

**Comment by glouppe** [comment]
@MechCoder @betatim I have made the necessary changes to have a full API representing the original and transformed space. Next step is now to change all `*_minimize` functions. I'll start that tonight or tomorrow.

**Comment by MechCoder** [comment]
Thanks! The new interface looks great. I'll give a detailed review tonight.

Also, do you think we should merge https://github.com/scikit-optimize/scikit-optimize/pull/62 and https://github.com/scikit-learn/scikit-learn/issues/6868 (at least port this issue) before refactoring `gp_minimize`?

**Comment by glouppe** [comment]
Too late, I ported `gp_minimize` already, in 246339ce01898c92ce9e143895c5def9c46dcb78.

**Comment by glouppe** [comment]
Bim, job done! Everything should have been updated. Ready for reviews :)

File: skopt/space.py [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+        rows = []
+
+        for i in range(n_samples):
+            r = []
+            for j in range(self.n_dims):
+                r.append(columns[j][i])
+
+            rows.append(r)
+
+        return np.asarray(rows)
```
Comment (MechCoder) [inline skopt/space.py]: Doing this creates object dtypes. What is your opinion about just returning a list of lists?

File: skopt/space.py [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+            The transformed samples.
+        """
+        # Pack by dimension
+        columns = []
+        for dim in self.dimensions:
+            columns.append([])
+
+        for i in range(len(X)):
+            for j in range(self.n_dims):
+                columns[j].append(X[i][j])
```
Comment (MechCoder) [inline skopt/space.py]: You can do `columns[j].append(self.dimensions[j].transform(X[i][j]))` here...

Comment (MechCoder) [inline skopt/space.py]: Do you think it is worthy to have a `dtype` proprty for each `Distribution` object? If we do that we can simply do:

```python
transformed_columns = []
for i, dim in enumerate(self.dimensions):
    curr_dtype = dim.dtype
    transformed_columns.append(dim.transform(dim.columns[:, i].astype(curr_dtype)).reshape(len(X), -1))

return np.hstack(transformed_columns)
```

File: skopt/space.py [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+        rows = []
+
+        for i in range(len(Xt)):
+            r = []
+            for j in range(self.n_dims):
+                r.append(columns[j][i])
+
+            rows.append(r)
+
+        return np.asarray(rows)
```
Comment (MechCoder) [inline skopt/space.py]: `np.transpose(columns)` does not work? (Neglecting the fact that it will return an object dtype)

File: skopt/space.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+
+    @property
+    def n_dims(self):
+        """The dimensionality of the original space."""
+        return sum([dim.size for dim in self.dimensions])
```
Comment (MechCoder) [inline skopt/space.py]: This is just `len(self.dimensions)` right?

Comment (glouppe) [inline skopt/space.py]: Assuming that dimensions are all of size 1, yes. (I think it is a safe assumption, I'll change that.)

File: skopt/space.py (hunk 1 above)
Comment (glouppe) [inline skopt/space.py]: It should create object dtypes only when there is Categorical variables in the mix right? I like returning a Numpy array, as it often makes it easier then to write the function to optimize. I.e., we would have to rewrite the benchmarks functions so that they accept lists. Maybe we could do that in another PR?

File: skopt/space.py [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+
+    @property
+    def transformed_bounds(self):
+        if self.prior == "uniform":
+            return (self._low, self._high)
+
+        elif self.prior == "log-uniform":
+            return (np.log10(self._low), np.log10(self._high))
+
+        else:
```
Comment (MechCoder) [inline skopt/space.py]: nit: this will never happen right?

File: skopt/space.py [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+    def transformed_size(self):
+        return len(self.categories)
+
+    @property
+    def bounds(self):
+        return self.categories
+
+    @property
+    def transformed_bounds(self):
+        return [(0.0, 1.0) for i in range(len(self.categories))]
```
Comment (MechCoder) [inline skopt/space.py]: np.tile?

File: skopt/space.py (hunk 1 above)
Comment (MechCoder) [inline skopt/space.py]: Yes, I agree. I looked at this file first and hence assumed that the `_optimize` functions had been refactored to accept categorical parameters.

File: skopt/space.py (hunk 3 above)
Comment (MechCoder) [inline skopt/space.py]: `np.transpose(columns)` should work here as well..

**Comment by MechCoder** [comment]
Just some minor comments, I'll give another pass tomorrow and that should be it.

Sleepy now :yawn:

**Comment by betatim** [comment]
Do we still need [this hack](https://github.com/scikit-optimize/scikit-optimize/pull/83/files#diff-89203a9ab976fbdea06be9034e00772bL30)?

File: skopt/dummy_opt.py [hunk 7]
```diff
@@ -15,9 +15,17 @@ def dummy_minimize(func, bounds, maxiter=1000, random_state=None):
...
-        - ``bounds[i][1]`` should give the upper bound of each parameter.
+    * `dimensions` [list, shape=(n_dims,)]:
+        List of search space dimensions.
+        Each search dimension can be defined either as
+
+        - a `(upper_bound, lower_bound)` tuple (for `Real` or `Integer`
+          dimensions),
+        - a `(upper_bound, lower_bound, "prior")` tuple (for `Real`
+          dimensions),
+        - an instance of a `Dimension` object (for `Real`, `Integer` or
```
Comment (betatim) [inline skopt/dummy_opt.py]: I'd swap lines 26 and 28 and change to "an instance of a `Dimension` object (`Real`, `Integer` or `Categorical`)" (remove the for)

File: skopt/dummy_opt.py [hunk 8]
```diff
@@ -43,16 +51,14 @@ def dummy_minimize(func, bounds, maxiter=1000, random_state=None):
...
-    n_params = len(bounds)
-    lb, ub = extract_bounds(bounds)
-
-    X = lb + (ub - lb) * rng.rand(maxiter, n_params)
     init_y = func(X[0])
     if not np.isscalar(init_y):
         raise ValueError(
             "The function to be optimized should return a scalar")
-    y = np.asarray([init_y] + [func(X[i]) for i in range(maxiter - 1)])
+    y = np.asarray([init_y] + [func(X[1 + i]) for i in range(maxiter - 1)])
```
Comment (betatim) [inline skopt/dummy_opt.py]: How about `[func(X[i]) for i in range(1, maxiter)]` to get rid of the +1 -1 business?

File: skopt/gbrt_opt.py [hunk 9]
```diff
@@ -37,9 +28,17 @@ def gbrt_minimize(func, bounds, base_estimator=None, maxiter=100,
...
-        - ``bounds[i][1]`` should give the upper bound of each parameter.
+    * `dimensions` [list, shape=(n_dims,)]:
+        List of search space dimensions.
+        Each search dimension can be defined either as
+
+        - a `(upper_bound, lower_bound)` tuple (for `Real` or `Integer`
+          dimensions),
+        - a `(upper_bound, lower_bound, "prior")` tuple (for `Real`
+          dimensions),
+        - an instance of a `Dimension` object (for `Real`, `Integer` or
```
Comment (betatim) [inline skopt/gbrt_opt.py]: same as above

File: skopt/gp_opt.py [hunk 10]
```diff
@@ -47,9 +47,17 @@ def gp_minimize(func, bounds, base_estimator=None, acq="LCB", xi=0.01,
...
-        - ``bounds[i][1]`` should give the upper bound of each parameter.
+    * `dimensions` [list, shape=(n_dims,)]:
+        List of search space dimensions.
+        Each search dimension can be defined either as
+
+        - a `(upper_bound, lower_bound)` tuple (for `Real` or `Integer`
+          dimensions),
+        - a `(upper_bound, lower_bound, "prior")` tuple (for `Real`
+          dimensions),
+        - an instance of a `Dimension` object (for `Real`, `Integer` or
```
Comment (betatim) [inline skopt/gp_opt.py]: same as above

File: skopt/space.py [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+        self._low = low
+        self._high = high
+        self._rvs = randint(self._low, self._high + 1)
+        self.transformer = _Identity()
+
+    def inverse_transform(self, Xt):
+        """Inverse transform samples from the warped space back into the
+           orignal space.
+        """
+        return super(Integer, self).inverse_transform(Xt).astype(np.int)
```
Comment (betatim) [inline skopt/space.py]: I can't work out why we need this, can you add a comment that explains it to future us?

File: skopt/space.py [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+            are equally likely.
+        """
+        self.categories = np.asarray(categories)
+        self.transformer = _CategoricalEncoder()
+        self.transformer.fit(self.categories)
+
+        if prior is None:
+            prior = np.tile(1. / len(self.categories), len(self.categories))
+
+        self._rvs = rv_discrete(values=(range(len(self.categories)), prior))
```
Comment (betatim) [inline skopt/space.py]: We should check that `sum(prior)==1`, `rv_discrete` doesn't seem to do that

File: skopt/space.py [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+            else:
+                columns.append(dim.rvs(n_samples=n_samples, random_state=rng))
+
+        return np.transpose(columns)
+
+    def transform(self, X):
+        """Transform samples from the original space into a warped space.
+
+        Note: this transformation is expected to be used to project samples
+              in a space where numerical optimization is suited.
```
Comment (betatim) [inline skopt/space.py]: -> "into a suitable space for numerical optimization."

**Comment by betatim** [comment]
No additional comments.

**Comment by glouppe** [comment]
Thanks both of you for comments! I think I addressed most of them. I'll merge if Travis is happy, so that we can all keep iterating over this.

**Comment by MechCoder** [comment]
@glouppe Thanks! I"ve addressed a minor optimization here (https://github.com/scikit-optimize/scikit-optimize/pull/85)
